include/bout/petsc_interface.hxx

```diff
@@ -0,0 +1,576 @@
+  Mesh* fieldmesh;
+
+private:
+  /// This gets called by initialiseTest and is used to register
+  /// fields with fake parallel meshes.
+  virtual void registerFieldForTest(FieldData& f);
+  virtual void registerFieldForTest(FieldPerp& f);
+
+  PetscLib lib;
+
+  /// Fields containing the indices for each element (as reals)
+  Field3D indices3D;
+  Field2D indices2D;
+  FieldPerp indicesPerp;
+  
+  /// The only instance of this class acting on the global Mesh
+  static IndexerPtr globalInstance;
+  static bool initialisedGlobal;
+  static Mesh* globalmesh;
+  bool initialised;
+};
+
+
+/*!
+ * A class which is used to assign to a particular element of a PETSc
+ * vector. It is meant to be transient and will be destroyed immediately
+ * after use. In general you should not try to assign an instance to a
+ * variable.
+ */
+class PetscVectorElement {
+public:
+  BoutReal operator=(BoutReal val);
+  BoutReal operator+=(BoutReal val);
+
+  /// This is the only valid method for constructing new instances,
+  /// guaranteeing they can be safely deleted once used.
+  static PetscVectorElement& newElement(Vec* vector, PetscInt index);
+
+  // Prevent non-transient copies from being created
+  PetscVectorElement() = delete;
+  PetscVectorElement(const PetscVectorElement& p) = delete;
+  PetscVectorElement& operator=(const PetscVectorElement& rhs) = delete;
+  PetscVectorElement& operator=(PetscVectorElement&& rhs) = delete;
+
+private:
+  PetscVectorElement(Vec* vector, int index);
+  Vec* petscVector;
+  PetscInt petscIndex;
+};
+
+
+/*!
+ * A class which is used to assign to a particular element of a PETSc
+ * matrix, potentially with a y-offset. It is meant to be transient
+ * and will be destroyed immediately after use. In general you should
+ * not try to assign an instance to a variable.
+ */
+class PetscMatrixElement {
+public:
+  BoutReal operator=(BoutReal val);
+  BoutReal operator+=(BoutReal val);
+
+  /// This is the only valid method for constructing new instances,
+  /// guaranteeing they can be safely deleted once used.
+  static PetscMatrixElement& newElement(Mat* matrix, PetscInt row, PetscInt col,
+					std::vector<PetscInt> p = std::vector<PetscInt>(),
+					std::vector<BoutReal> w = std::vector<BoutReal>());
+
+  // Prevent non-transient copies from being created
+  PetscMatrixElement() = delete;
+  PetscMatrixElement(const PetscMatrixElement& p) = delete;
+  PetscMatrixElement& operator=(const PetscMatrixElement& rhs) = delete;
+  PetscMatrixElement& operator=(PetscMatrixElement&& rhs) = delete;
+
+private:
+  PetscMatrixElement(Mat* matrix, PetscInt row,
+		     std::vector<PetscInt> p, std::vector<BoutReal> w);
+  void setValues(BoutReal val, InsertMode mode);
+  Mat* petscMatrix;
+  PetscInt petscRow;
+  std::vector<PetscInt> positions;
+  std::vector<BoutReal> weights;
+};
+
+
+/*!
+ * A class which wraps PETSc vector objects, allowing them to be
+ * indexed using the BOUT++ scheme.
+ */
+template<class F> class PetscVector;
```

Comment:
Just style, but generally prefer `T` for template arguments

Reply:
I was using F to indicate it was a field rather than any old type, but I've now changed it.